# **Data cleaning and exploration**

This notebook is used to clean the column of listings dataset and explore.

### 1. Import packages and data set

In [3]:
# Loading packages
import pandas as pd
import glob

# Other
import warnings
warnings.filterwarnings(action="ignore")
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# setting
pd.set_option('display.max_columns', None)  # or 1000
pd.set_option('display.max_rows', None)  # or 1000
# pd.set_option('display.max_colwidth', -1)  # or 199 - warning
pd.options.display.max_colwidth = 100

### 2. check sample dataset

In [4]:
# Random check for csv files
df_sample = pd.read_csv('listings/01_2021.csv').rename(columns=str.lower)
df_sample.head(3)

id                         listing_url       scrape_id last_scraped  \
0  11156  https://www.airbnb.com/rooms/11156  20210110040047   2021-01-13   
1  12351  https://www.airbnb.com/rooms/12351  20210110040047   2021-01-11   
2  14250  https://www.airbnb.com/rooms/14250  20210110040047   2021-01-13   

                                name  \
0               An Oasis in the City   
1  Sydney City & Harbour at the door   
2                Manly Harbour House   

                                                                                           description  \
0  Very central to the city which can be reached by an easy walk or by bus, with transport at the d...   
1  Come stay with Vinh & Stuart (Awarded as one of Australia's top hosts by Airbnb CEO Brian Chesky...   
2  Beautifully renovated, spacious and quiet, our 3 Bedroom, 3 Bathroom home is only a  10 minute w...   

                                                                                 neighborhood_overview  \
0  It is very close to everything and everywhere, has  many trees along the streets and is in a bea...   
1  Pyrmont is an inner-city village of Sydney, only about 2kms from the Sydney CBD (Central Busines...   
2  Balgowlah Heights is one of the most prestigious areas on the Northern Beaches.  Filled with sea...   

                                                                 picture_url  \
0             https://a0.muscache.com/pictures/2797669/17895d03_original.jpg   
1  https://a0.muscache.com/pictures/763ad5c8-c951-43e0-b926-4a98c25c45e8.jpg   
2            https://a0.muscache.com/pictures/56935671/fdb8c0bf_original.jpg   

   host_id                                 host_url host_name  host_since  \
0    40855  https://www.airbnb.com/users/show/40855   Colleen  2009-09-23   
1    17061  https://www.airbnb.com/users/show/17061    Stuart  2009-05-14   
2    55948  https://www.airbnb.com/users/show/55948     Heidi  2009-11-20   

                             host_location  \
0  Potts Point, New South Wales, Australia   
1       Sydney, New South Wales, Australia   
2       Sydney, New South Wales, Australia   

                                                                                            host_about  \
0  Recently retired, I've lived & worked on 4 continents & over the years have travelled thru about...   
1  G'Day from Australia!\r\n\r\nHe's Vinh, and I'm Stuart...\r\n\r\nWe're big fans of AirBnB for ne...   
2  I am a Canadian who has made Australia her home for over ten years. I appreciate a good place to...   

  host_response_time host_response_rate host_acceptance_rate  \
0                NaN                NaN                 100%   
1                NaN                NaN                  NaN   
2     within an hour               100%                  73%   

  host_is_superhost  \
0                 f   
1                 f   
2                 t   

                                                                                    host_thumbnail_url  \
0  https://a0.muscache.com/im/users/40855/profile_pic/1259121939/original.jpg?aki_policy=profile_small   
1  https://a0.muscache.com/im/users/17061/profile_pic/1425534501/original.jpg?aki_policy=profile_small   
2  https://a0.muscache.com/im/users/55948/profile_pic/1327370146/original.jpg?aki_policy=profile_small   

                                                                                      host_picture_url  \
0  https://a0.muscache.com/im/users/40855/profile_pic/1259121939/original.jpg?aki_policy=profile_x_...   
1  https://a0.muscache.com/im/users/17061/profile_pic/1425534501/original.jpg?aki_policy=profile_x_...   
2  https://a0.muscache.com/im/users/55948/profile_pic/1327370146/original.jpg?aki_policy=profile_x_...   

  host_neighbourhood  host_listings_count  host_total_listings_count  \
0        Potts Point                  1.0                        1.0   
1            Pyrmont                  2.0                        2.0   
2          Balgowlah          

### 2. check the datasets in listings folder

In [5]:
glob.glob("listings/*.csv")

['listings/03_2021.csv',
 'listings/02_2021.csv',
 'listings/08_2020.csv',
 'listings/09_2020.csv',
 'listings/04_2021.csv',
 'listings/05_2020.csv',
 'listings/12_2020.csv',
 'listings/01_2021.csv',
 'listings/11_2020.csv',
 'listings/10_2020.csv',
 'listings/06_2020.csv',
 'listings/07_2020.csv']

In [6]:
# Check for csv dimensions
files = glob.glob("listings/*.csv")
print("Number of listings file:", len(files))
for file in files:
    print(file, ": ", pd.read_csv(file).shape)

Number of listings file: 12
listings/03_2021.csv :  (33229, 74)
listings/02_2021.csv :  (33630, 74)
listings/08_2020.csv :  (31391, 74)
listings/09_2020.csv :  (34829, 74)
listings/04_2021.csv :  (32679, 74)
listings/05_2020.csv :  (37562, 106)
listings/12_2020.csv :  (33871, 74)
listings/01_2021.csv :  (33902, 74)
listings/11_2020.csv :  (33795, 74)
listings/10_2020.csv :  (34276, 74)
listings/06_2020.csv :  (36901, 106)
listings/07_2020.csv :  (36057, 102)


**Note:**
1. DtypeWarning here, need to be addressed in later part
2. No. of columns are different, need to find out all the common columns for ELT pipeline

There are mixed types in some columns in below files:    
* listings/05_2020.csv : Columns (43,61,62,94) have mixed types.
* listings/06_2020.csv :  Columns (43,61,62,94) have mixed types
* listings/07_2020.csv :  (36057, 102) Columns (49,58,59,89) have mixed types.

### 4. Find common columns

In [7]:
cols = []

for file in files:
    current_col_list = pd.read_csv(file).columns.tolist()   
    cols.append(current_col_list)
    
common_cols = set(cols[0]).intersection(*cols)
print("Number of common columns:", len(common_cols))
print(common_cols)

Number of common columns: 67
{'host_acceptance_rate', 'room_type', 'calculated_host_listings_count_shared_rooms', 'property_type', 'review_scores_accuracy', 'host_picture_url', 'host_neighbourhood', 'listing_url', 'host_identity_verified', 'reviews_per_month', 'availability_90', 'has_availability', 'price', 'accommodates', 'calculated_host_listings_count', 'neighbourhood', 'scrape_id', 'minimum_minimum_nights', 'host_total_listings_count', 'description', 'host_about', 'calendar_updated', 'host_is_superhost', 'neighbourhood_group_cleansed', 'instant_bookable', 'host_name', 'host_listings_count', 'availability_365', 'host_thumbnail_url', 'host_has_profile_pic', 'maximum_nights_avg_ntm', 'number_of_reviews_ltm', 'review_scores_communication', 'license', 'host_since', 'latitude', 'review_scores_value', 'review_scores_cleanliness', 'calculated_host_listings_count_private_rooms', 'name', 'maximum_minimum_nights', 'number_of_reviews', 'id', 'host_url', 'availability_60', 'neighborhood_overvie

In [8]:
for file in files:
    print(file)
    col = pd.read_csv(file).columns.tolist()
    print([x for x in col if x not in common_cols])
    print()

listings/03_2021.csv
['bathrooms_text', 'bedrooms', 'beds', 'amenities', 'minimum_nights', 'maximum_nights', 'number_of_reviews_l30d']

listings/02_2021.csv
['bathrooms_text', 'bedrooms', 'beds', 'amenities', 'minimum_nights', 'maximum_nights', 'number_of_reviews_l30d']

listings/08_2020.csv
['bathrooms_text', 'bedrooms', 'beds', 'amenities', 'minimum_nights', 'maximum_nights', 'number_of_reviews_l30d']

listings/09_2020.csv
['bathrooms_text', 'bedrooms', 'beds', 'amenities', 'minimum_nights', 'maximum_nights', 'number_of_reviews_l30d']

listings/04_2021.csv
['bathrooms_text', 'bedrooms', 'beds', 'amenities', 'minimum_nights', 'maximum_nights', 'number_of_reviews_l30d']

listings/05_2020.csv
['summary', 'space', 'experiences_offered', 'notes', 'transit', 'access', 'interaction', 'house_rules', 'thumbnail_url', 'medium_url', 'xl_picture_url', 'street', 'city', 'state', 'zipcode', 'market', 'smart_location', 'country_code', 'country', 'is_location_exact', 'bedrooms', 'beds', 'bed_type', 

In [11]:
# for an airbnb dataset, it is reasonable to include:
more_cols = ['bathrooms_text', 'bedrooms', 'beds', 'amenities', 'minimum_nights', 'maximum_nights', 'number_of_reviews_l30d']
          
for file in files:
    print(file)
    col = pd.read_csv(file).columns.tolist()
    print([x for x in more_cols if x not in col])
    print()

listings/03_2021.csv
[]

listings/02_2021.csv
[]

listings/08_2020.csv
[]

listings/09_2020.csv
[]

listings/04_2021.csv
[]

listings/05_2020.csv
['bathrooms_text', 'number_of_reviews_l30d']

listings/12_2020.csv
[]

listings/01_2021.csv
[]

listings/11_2020.csv
[]

listings/10_2020.csv
[]

listings/06_2020.csv
['bathrooms_text', 'number_of_reviews_l30d']

listings/07_2020.csv
['bathrooms_text', 'bedrooms', 'beds', 'amenities', 'minimum_nights', 'maximum_nights']



csv for May, June and July need to fill these columns with null in snowflake, not giving up these columns because of 1 file missing

In [13]:
common_cols = list(common_cols)
cols = common_cols + more_cols
len(cols)

74

In [41]:
new_cols = ['id', 'listing_url', 'scrape_id', 'last_scraped', 'name', 'description', 'neighborhood_overview', 'picture_url', 
            'host_id', 'host_url', 'host_name', 'host_since', 'host_location', 'host_about', 'host_response_time', 
            'host_response_rate', 'host_acceptance_rate', 'host_is_superhost', 'host_thumbnail_url', 'host_picture_url', 
            'host_neighbourhood', 'host_listings_count', 'host_total_listings_count', 'host_verifications', 'host_has_profile_pic', 
            'host_identity_verified', 'neighbourhood', 'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude', 
            'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms', 'bathrooms_text', 'bedrooms', 'beds', 
            'amenities', 'price', 'minimum_nights', 'maximum_nights', 'minimum_minimum_nights', 'maximum_minimum_nights', 
            'minimum_maximum_nights', 'maximum_maximum_nights', 'minimum_nights_avg_ntm', 'maximum_nights_avg_ntm', 
            'calendar_updated', 'has_availability', 'availability_30', 'availability_60', 'availability_90', 'availability_365', 
            'calendar_last_scraped', 'number_of_reviews', 'number_of_reviews_ltm', 'number_of_reviews_l30d', 'first_review', 
            'last_review', 'review_scores_rating', 'review_scores_accuracy', 'review_scores_cleanliness', 'review_scores_checkin', 
            'review_scores_communication', 'review_scores_location', 'review_scores_value', 'license', 'instant_bookable', 
            'calculated_host_listings_count', 'calculated_host_listings_count_entire_homes', 'calculated_host_listings_count_private_rooms', 
            'calculated_host_listings_count_shared_rooms', 'reviews_per_month']

In [35]:
#check for these columns
files = glob.glob("listings/*.csv")
for file in files:
    file_name = file.split(".")[0]
    save_file = file_name + "_new.csv"
    
    # read data
    df = pd.read_csv(file)
    
    # change 'last_scraped' to datetime format
    df['last_scraped'] = pd.to_datetime(df['last_scraped'])
    
    # fill na to csvs
    df_col = df.columns.tolist()
    for i in new_cols:
        if i not in df_col:
            df[i] = ''
    df_new = df[new_cols]
    df_new.to_csv(save_file, index=False)

In [23]:
# Check for csv dimensions
files = glob.glob("listings/*new.csv")
print("Number of listings file:", len(files))
for file in files:
    print(file, ": ", pd.read_csv(file).shape)

Number of listings file: 12
listings/05_2020_new.csv :  (37562, 74)
listings/10_2020_new.csv :  (34277, 74)
listings/03_2021_new.csv :  (33229, 74)
listings/04_2021_new.csv :  (32679, 74)
listings/11_2020_new.csv :  (33796, 74)
listings/02_2021_new.csv :  (33630, 74)
listings/09_2020_new.csv :  (34830, 74)
listings/08_2020_new.csv :  (31391, 74)
listings/06_2020_new.csv :  (36901, 74)
listings/07_2020_new.csv :  (36057, 74)
listings/12_2020_new.csv :  (33871, 74)
listings/01_2021_new.csv :  (33902, 74)
